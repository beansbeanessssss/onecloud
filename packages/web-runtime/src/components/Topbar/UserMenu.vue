<template>
  <nav :aria-label="$gettext('Account menu')">
    <oc-button
      id="_userMenuButton"
      ref="menuButton"
      v-oc-tooltip="$gettext('My Account')"
      class="oc-topbar-personal"
      appearance="raw"
      :aria-label="$gettext('My Account')"
    >
      <avatar-image
        v-if="onPremisesSamAccountName"
        class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"
        :width="32"
        :userid="onPremisesSamAccountName"
        :user-name="user.displayName"
      />
      <oc-avatar-item
        v-else
        class="oc-topbar-unauthenticated-avatar oc-flex-inline oc-flex-center oc-flex-middle"
        :name="$gettext('User Menu login')"
        :width="32"
        icon="user-add"
        icon-fill-type="line"
        icon-color="var(--oc-color-swatch-brand-default)"
        background="var(--oc-color-swatch-brand-contrast)"
      />
    </oc-button>
    <oc-drop
      ref="menu"
      drop-id="account-info-container"
      toggle="#_userMenuButton"
      mode="click"
      close-on-click
      padding-size="small"
      class="oc-overflow-hidden"
    >
      <oc-list class="user-menu-list">
        <template v-if="!onPremisesSamAccountName">
          <li>
            <oc-button
              id="oc-topbar-account-login"
              appearance="raw"
              type="router-link"
              :to="loginLink"
            >
              <oc-icon name="login-box" fill-type="line" class="oc-p-xs" />
              <span v-text="$gettext('Log in')" />
            </oc-button>
          </li>
        </template>
        <template v-else>
          <li>
            <oc-button
              id="oc-topbar-account-manage"
              type="router-link"
              :to="{ path: '/account' }"
              appearance="raw"
            >
              <avatar-image
                :width="32"
                :userid="onPremisesSamAccountName"
                :user-name="user.displayName"
              />
              <span class="profile-info-wrapper" :class="{ 'oc-py-xs': !user.mail }">
                <span class="oc-display-block" v-text="user.displayName" />
                <span v-if="user.mail" class="oc-text-small" v-text="user.mail" />
              </span>
            </oc-button>
          </li>
          <li v-for="(app, index) in applicationsList" :key="`user-menu-${index}`">
            <oc-button
              v-if="app.url"
              type="a"
              appearance="raw"
              :target="app.target"
              :href="app.url"
            >
              <oc-icon :name="app.icon" class="oc-p-xs" />
              <span v-text="$gettext(app.title)" />
            </oc-button>
            <oc-button v-else type="router-link" appearance="raw" :to="{ path: app.path }">
              <oc-icon :name="app.icon" class="oc-p-xs" />
              <span v-text="$gettext(app.title)" />
            </oc-button>
          </li>
          <li v-if="quotaEnabled" class="storage-wrapper oc-pl-s">
            <oc-icon name="cloud" fill-type="line" class="oc-p-xs" />
            <div class="storage-wrapper-text oc-width-1-1">
              <p class="oc-my-rm">
                <span class="oc-display-block" v-text="personalStorageLabel" />
                <span class="oc-text-small" v-text="personalStorageDetailsLabel" />
              </p>
              <oc-progress
                v-if="limitedPersonalStorage"
                :value="quotaUsagePercent"
                :max="100"
                size="small"
                :variation="quotaProgressVariant"
              />
            </div>
          </li>
          <li>
            <oc-button id="oc-topbar-account-logout" appearance="raw" @click="logout">
              <oc-icon name="logout-box-r" fill-type="line" class="oc-p-xs" />
              <span v-text="$gettext('Log out')" />
            </oc-button>
          </li>
        </template>
      </oc-list>
      <div v-if="imprintUrl || privacyUrl" class="imprint-footer oc-py-s oc-mt-m oc-text-center">
        <oc-button v-if="imprintUrl" type="a" appearance="raw" :href="imprintUrl" target="_blank"
          ><span v-text="$gettext('Imprint')"
        /></oc-button>
        <span v-if="privacyUrl">Â·</span>
        <oc-button v-if="privacyUrl" type="a" appearance="raw" :href="privacyUrl" target="_blank"
          ><span v-text="$gettext('Privacy')"
        /></oc-button>
      </div>
    </oc-drop>
  </nav>
</template>

<script lang="ts">
import { storeToRefs } from 'pinia'
import { defineComponent, PropType, ComponentPublicInstance, computed, unref } from 'vue'
import filesize from 'filesize'
import { authService } from '../../services/auth'
import { useGetMatchingSpace, useRoute, useThemeStore, useUserStore } from '@ownclouders/web-pkg'
import { OcDrop } from 'design-system/src/components'

export default defineComponent({
  props: {
    applicationsList: {
      type: Array as PropType<any>,
      required: false,
      default: () => []
    }
  },
  setup() {
    const route = useRoute()
    const userStore = useUserStore()
    const themeStore = useThemeStore()
    const { getPersonalSpace } = useGetMatchingSpace()

    const { user } = storeToRefs(userStore)

    const loginLink = computed(() => {
      return {
        name: 'login',
        query: { redirectUrl: unref(route).fullPath }
      }
    })

    const imprintUrl = computed(() => themeStore.currentTheme.common.urls.imprint)
    const privacyUrl = computed(() => themeStore.currentTheme.common.urls.privacy)

    const quota = computed(() => {
      return getPersonalSpace()?.spaceQuota
    })

    return {
      user,
      loginLink,
      imprintUrl,
      privacyUrl,
      quota
    }
  },
  computed: {
    onPremisesSamAccountName() {
      return this.user?.onPremisesSamAccountName
    },
    personalStorageLabel() {
      if (!this.limitedPersonalStorage) {
        return this.$gettext('Personal storage')
      }
      return this.$gettext('Personal storage (%{percentage}% used)', {
        percentage: (this.quotaUsagePercent || 0).toString()
      })
    },
    personalStorageDetailsLabel() {
      const total = this.quota.definition === 'none' ? 0 : this.quota.total || 0
      const used = this.quota.used || 0
      return total
        ? this.$gettext('%{used} of %{total} used', {
            used: filesize(used),
            total: filesize(total)
          })
        : this.$gettext('%{used} used', {
            used: filesize(used),
            total: filesize(total)
          })
    },
    limitedPersonalStorage() {
      return this.quota.total !== 0
    },
    quotaEnabled() {
      return !!this.quota
    },
    quotaUsagePercent() {
      return parseFloat(((this.quota.used / this.quota.total) * 100).toFixed(2))
    },

    quotaProgressVariant() {
      if ((this.quotaUsagePercent || 0) < 80) {
        return 'primary'
      }
      if ((this.quotaUsagePercent || 0) < 90) {
        return 'warning'
      }
      return 'danger'
    }
  },
  mounted() {
    ;(this.$refs.menu as InstanceType<typeof OcDrop>)?.tippy?.setProps({
      onHidden: () => (this.$refs.menuButton as ComponentPublicInstance).$el.focus(),
      onShown: () =>
        (this.$refs.menu as ComponentPublicInstance).$el.querySelector('a:first-of-type').focus()
    })
  },
  methods: {
    logout() {
      authService.logoutUser()
    }
  }
})
</script>

<style lang="scss" scoped>
.user-menu-list li {
  align-items: center;
  display: flex;
  margin: var(--oc-space-xsmall) 0;

  &:first-child {
    margin-top: 0;
  }

  &:last-child {
    margin-bottom: 0;
  }

  a,
  button {
    gap: var(--oc-space-medium);
    justify-content: flex-start;
    min-height: 3rem;
    padding-left: var(--oc-space-small);
    width: 100%;

    &:focus,
    &:hover {
      background-color: var(--oc-color-background-hover);
      color: var(--oc-color-swatch-passive-default);
      text-decoration: none;
    }

    .profile-info-wrapper {
      text-align: left;
      word-break: break-all;
      line-height: initial;
    }
  }

  &.storage-wrapper {
    gap: var(--oc-space-medium);
    min-height: 3rem;

    .storage-wrapper-text {
      align-self: flex-end;
    }
  }
}

.imprint-footer {
  background-color: var(--oc-color-background-hover);
  margin-left: calc(var(--oc-space-small) * -1);
  width: calc(100% + var(--oc-space-small) * 2);
  margin-bottom: calc(var(--oc-space-small) * -1) !important;

  a {
    font-size: var(--oc-font-size-medium) !important;
    color: var(--oc-color-text-default);
  }
}
</style>
