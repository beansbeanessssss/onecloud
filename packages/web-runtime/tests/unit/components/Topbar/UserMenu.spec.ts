import UserMenu from 'web-runtime/src/components/Topbar/UserMenu.vue'
import {
  createStore,
  defaultPlugins,
  defaultStubs,
  mount,
  defaultStoreMockOptions,
  defaultComponentMocks,
  RouteLocation
} from 'web-test-helpers'
import { mock } from 'jest-mock-extended'

const totalQuota = 1000
const basicQuota = 300
const warningQuota = 810
const dangerQuota = 910

const basicRelativeQuota = (basicQuota / totalQuota) * 100
const warningRelativeQuota = (warningQuota / totalQuota) * 100
const dangerRelativeQuota = (dangerQuota / totalQuota) * 100

const noEmail = ''
const email = 'test@test.de'

describe('User Menu component', () => {
  describe('when user is not logged in', () => {
    it('renders a navigation with login button', () => {
      const wrapper = getMountedWrapper({}, noEmail, true)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota and no email is set', () => {
    it('renders a navigation without email', () => {
      const wrapper = getMountedWrapper(
        { used: basicQuota, total: totalQuota, relative: basicRelativeQuota },
        noEmail
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when no quota and email is set', () => {
    it('the user menu does not contain a quota', () => {
      const wrapper = getMountedWrapper(null, email)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when no quota and no email is set', () => {
    it('the user menu does not contain a quota', () => {
      const wrapper = getMountedWrapper(null, noEmail)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota is below 80%', () => {
    it('renders a primary quota progress bar', () => {
      const wrapper = getMountedWrapper(
        {
          used: basicQuota,
          total: totalQuota,
          relative: basicRelativeQuota,
          definition: 'default'
        },
        email
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota is above 80% and below 90%', () => {
    it('renders a warning quota progress bar', () => {
      const wrapper = getMountedWrapper(
        {
          used: warningQuota,
          total: totalQuota,
          relative: warningRelativeQuota,
          definition: 'default'
        },
        email
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota is above 90%', () => {
    it('renders a danger quota progress bar', () => {
      const wrapper = getMountedWrapper(
        {
          used: dangerQuota,
          total: totalQuota,
          relative: dangerRelativeQuota,
          definition: 'default'
        },
        email
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota is unlimited', () => {
    it('renders no percentag of total and no progress bar', () => {
      const wrapper = getMountedWrapper(
        {
          used: basicQuota,
          definition: 'default'
        },
        email
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('when quota is not defined', () => {
    it('renders no percentag of total and no progress bar', () => {
      const wrapper = getMountedWrapper(
        {
          used: dangerQuota,
          total: totalQuota,
          relative: dangerRelativeQuota,
          definition: 'none'
        },
        email
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('imprint and privacy urls', () => {
    it('should renders imprint and privacy section if urls are defined', () => {
      const wrapper = getMountedWrapper(
        {
          used: dangerQuota,
          total: totalQuota,
          relative: dangerRelativeQuota,
          definition: 'none'
        },
        email,
        false,
        'https://imprint.url',
        'https://privacy.url'
      )
      const element = wrapper.find('.imprint-footer')
      expect(element.exists()).toBeTruthy()
      const output = element.html()
      expect(output).toContain('https://imprint.url')
      expect(output).toContain('https://privacy.url')
    })
    it('should use theme url values over config url values', () => {
      const wrapper = getMountedWrapper(
        {
          used: dangerQuota,
          total: totalQuota,
          relative: dangerRelativeQuota,
          definition: 'none'
        },
        email,
        false,
        'https://imprint.url.theme',
        'https://privacy.url.theme',
        'https://imprint.url.config',
        'https://privacy.url.config'
      )
      const element = wrapper.find('.imprint-footer')
      const output = element.html()
      expect(output).toContain('https://imprint.url.theme')
      expect(output).toContain('https://privacy.url.theme')
    })
    it('should use config url values as fallback', () => {
      const wrapper = getMountedWrapper(
        {
          used: dangerQuota,
          total: totalQuota,
          relative: dangerRelativeQuota,
          definition: 'none'
        },
        email,
        false,
        '',
        '',
        'https://imprint.url.config',
        'https://privacy.url.config'
      )
      const element = wrapper.find('.imprint-footer')
      const output = element.html()
      expect(output).toContain('https://imprint.url.config')
      expect(output).toContain('https://privacy.url.config')
    })
  })
})

const getMountedWrapper = (
  quota,
  userEmail,
  noUser = false,
  imprintUrlTheme = '',
  privacyUrlTheme = '',
  imprintUrlConfig = '',
  privacyUrlConfig = ''
) => {
  const mocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ path: '/files', fullPath: '/files' })
    })
  }
  const storeOptions = defaultStoreMockOptions
  storeOptions.getters.configuration.mockImplementation(() => ({
    currentTheme: {
      general: {
        imprintUrl: imprintUrlTheme,
        privacyUrl: privacyUrlTheme
      }
    },
    options: {
      imprintUrl: imprintUrlConfig,
      privacyUrl: privacyUrlConfig
    }
  }))

  storeOptions.getters.quota.mockImplementation(() => quota)
  storeOptions.getters.user.mockImplementation(() => {
    return noUser
      ? {}
      : {
          id: 'einstein',
          username: 'einstein',
          userDisplayName: 'Albert Einstein',
          userEmail
        }
  })
  const store = createStore(storeOptions)
  return mount(UserMenu, {
    props: {
      applicationsList: [
        {
          icon: 'application',
          path: '/settings',
          title: 'Settings'
        }
      ]
    },
    global: {
      renderStubDefaultSlot: true,
      plugins: [...defaultPlugins(), store],
      stubs: {
        ...defaultStubs,
        'oc-button': true,
        'oc-drop': true,
        'oc-list': true,
        'avatar-image': true,
        'oc-icon': true,
        'oc-progress': true
      },
      mocks
    }
  })
}
