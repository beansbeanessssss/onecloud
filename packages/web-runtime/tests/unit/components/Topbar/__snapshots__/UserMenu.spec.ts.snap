// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Menu component when no quota and email is set the user menu does not contain a quota 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <!--v-if-->
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when no quota and no email is set the user menu does not contain a quota 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <!--v-if-->
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota and no email is set renders a navigation without email 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage (30% used)</span>
                    <br>
                    <span class="oc-text-small">300 B of 1 kB used</span>
                  </p>
                  <oc-progress-stub max="100" size="small" value="30" variation="primary"></oc-progress-stub>
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota is above 80% and below 90% renders a warning quota progress bar 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage (81% used)</span>
                    <br>
                    <span class="oc-text-small">810 B of 1 kB used</span>
                  </p>
                  <oc-progress-stub max="100" size="small" value="81" variation="warning"></oc-progress-stub>
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota is above 90% renders a danger quota progress bar 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage (91% used)</span>
                    <br>
                    <span class="oc-text-small">910 B of 1 kB used</span>
                  </p>
                  <oc-progress-stub max="100" size="small" value="91" variation="danger"></oc-progress-stub>
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota is below 80% renders a primary quota progress bar 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage (30% used)</span>
                    <br>
                    <span class="oc-text-small">300 B of 1 kB used</span>
                  </p>
                  <oc-progress-stub max="100" size="small" value="30" variation="primary"></oc-progress-stub>
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota is not defined renders no percentag of total and no progress bar 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage</span>
                    <br>
                    <span class="oc-text-small">910 B used</span>
                  </p>
                  <!--v-if-->
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;

exports[`User Menu component when quota is unlimited renders no percentag of total and no progress bar 1`] = `
<nav aria-label="Account menu">
  <div>
    <div class="oc-expanding-dropdown">
      <div class="head">
        <oc-button-stub appearance="raw" class="dropdown-button" variation="inverse">
          <avatar-image-stub class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle" userid="einstein" width="32"></avatar-image-stub>
        </oc-button-stub>
      </div>
      <transition-stub appear="false" css="true" mode="out-in" name="drop-down-slide" persisted="true">
        <div class="dropdown-content" style="display: none; max-width: 0px; max-height: 0px;">
          <div style="width: 0px; height: 0px;">
            <!--v-if-->
            <oc-list-stub class="user-menu-list">
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-manage" to="[object Object]" type="router-link">
                  <avatar-image-stub userid="einstein" width="32"></avatar-image-stub>
                  <span class="profile-info-wrapper oc-py-xs">
                    <span></span>
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" to="[object Object]" type="router-link">
                  <oc-icon-stub class="oc-p-xs" name="application"></oc-icon-stub>
                  <span>Settings</span>
                </oc-button-stub>
              </li>
              <li>
                <oc-button-stub appearance="raw" id="oc-topbar-account-logout">
                  <oc-icon-stub class="oc-p-xs" fill-type="line" name="logout-box-r"></oc-icon-stub>
                  <span>Log out</span>
                </oc-button-stub>
              </li>
              <li class="storage-wrapper oc-pl-s">
                <oc-icon-stub class="oc-p-xs" fill-type="line" name="cloud"></oc-icon-stub>
                <div class="storage-wrapper-text">
                  <p class="oc-my-rm">
                    <span>Personal storage</span>
                    <br>
                    <span class="oc-text-small">300 B used</span>
                  </p>
                  <!--v-if-->
                </div>
              </li>
            </oc-list-stub>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</nav>
`;
